lib/ex_doc/retriever.ex

```diff
@@ -341,7 +341,9 @@ defmodule ExDoc.Retriever do
       annotations_for_docs.(metadata) ++
         callback_data.extra_annotations ++ annotations_from_metadata(metadata, module_metadata)
 
-    doc_ast = doc_ast(content_type, source_doc, file: doc_file, line: doc_line + 1)
+    doc_ast =
+      (source_doc && doc_ast(content_type, source_doc, file: doc_file, line: doc_line + 1)) ||
+        (callback_data[:doc_fallback] && callback_data.doc_fallback.())
```

Comment:
Apologies in case I am missing something obvious, but can `source_doc` ever be `nil`? However, even if `nil`, `doc_ast` would return `nil` anyway, so I think we can skip the `source_doc &&` bit. I would also add a small function, `doc_fallback(callback_data)` for the second part, so we end-up wit this:

```suggestion
      doc_ast(content_type, source_doc, file: doc_file, line: doc_line + 1) ||
        doc_fallback(callback_data)
```

WDYT?

Reply:
> Apologies in case I am missing something obvious, but can source_doc ever be nil? 

I was just doing the same thing as for function_data, might very well be that source_doc cannot be `nil`.

Refactor pushed.